apache/metron
# METRON-1677 UUIDv4 GUID is not Lucene friendly

With this change, when documents are written to Elasticsearch the document ID is no longer set as the Metron GUID, but instead left unset so that Elasticsearch can auto-generate it.  Doing this improves write performance into Elasticsearch.

This will also be the case for any Lucene based Indexer, including Solr.  This work only covers Elasticsearch, but the same should be done for Solr as part of a separate effort.

While the default table view looks the same, in the following screenshot I customized the table to show both the document ID and the GUID. 

This change is dependent on the following open pull requests.
- [ ] #1247
- [ ] #1254 
- [ ] #1259  

## Changes

* The `ElasticsearchRetrieveLatestDao` was updated since the GUID is no longer the document ID.  This instead does a terms query on the GUID field instead of an ID query.

* The `Document` class now contains an optional documentID field.  If the `Document` is retrieved from one of the DAOs this field will be populated.  When creating a new document, this field will be empty.

* Many of the integrations tests had to be updated because the GUID and document ID are now different.

* The Alert UI was updated so that it visually looks the same. By default, the Metron GUID is still shown as one of the first columns in the table.  

    * The table is actually showing the document's GUID field instead of the document ID as it was before.  The ID field remains, which contains the document ID generated by Elasticsearch.  The user can choose to add this to the table, if they like.


## Testing

1. Spin-up Full Dev.

1. Open up the Alerts UI and perform the following basic actions.
    * Search for alerts
    * Escalate an alert
    * Comment on an alert
    * Delete a comment from an alert
    * Create a meta-alert
    * Escalate a meta-alert

1. Click on the configure wheel and add the 'id' field to the table view.  This will now display both the GUID and document ID in the table. They of course will be different.

1.  Click on the 'guid' field in any row to filter the search results by the guid.
    ![screen shot 2018-11-14 at 2 44 21 pm](https://user-images.githubusercontent.com/2475409/48646597-53433300-e9b7-11e8-870b-f061af8cca47.png)

1. Click on the 'id' field to filter the search results by the document ID.
    ![screen shot 2018-11-14 at 2 44 08 pm](https://user-images.githubusercontent.com/2475409/48646566-3a3a8200-e9b7-11e8-8370-f596346d4a62.png)

1. Group by some fields to drill into the data.  In the tree view, click on the 'guid' column and ensure the data sorts correctly.  Do the same for the 'id' column that was added.
    ![screen shot 2018-11-14 at 2 46 43 pm](https://user-images.githubusercontent.com/2475409/48646519-1414e200-e9b7-11e8-96a2-50c568d909b2.png)

## Pull Request Checklist
